ak-lite: add rollback support

ricardosalveti opened the pull request:
Tested scenarios (on a rpi3-64):
1 - Update in daemon mode
2 - Update via update command (to a newer rev, after update applied and after reboot)
3 - Rollback process in daemon mode
4 - Forced update after a rollback (protection is only done in daemon mode)
5 - Base image rollback with with docker app
6 - Docker app updates after a rollback
7 - Docker app updates with update
8 - Docker app updates with daemon
9 - Base image + docker app updates

To test the rollback support you will have to flash an image produced by the https://github.com/foundriesio/lmp-manifest/pull/48 pull request (build 734 uses my aktualizr branch so no need to build it by hand). This is required because we also need u-boot script updates, which is not yet managed by ostree.

The easiest way to trigger a rollback is to call aktualizr-lite update and remove the new kernel from the /boot directory, as then u-boot will fail to boot the new image and rollback to the previous one, like:
```
aktualizr-lite --command update
# Check the lmp-hash used by looking at the first kernel_image entry at /boot/loader/uEnv.txt
cd /boot/
sudo mv ./ostree/lmp-1f3c7bbb8edc3c407c968565a8d26162cb617d07dbfe86f26795670193814d4f/vmlinuz ./ostree/lmp-1f3c7bbb8edc3c407c968565a8d26162cb617d07dbfe86f26795670193814d4f/vmlinuz-orig
sudo touch ./ostree/lmp-1f3c7bbb8edc3c407c968565a8d26162cb617d07dbfe86f26795670193814d4f/vmlinuz
```

Then if you start ak-lite in daemon mode you will see that it will refuse to update as the image is known to be bad. Forcing via update command will still work, but you might have to move the vmlinuz-orig back as it looks like a new ostree deployment won't refresh that file.

ricardosalveti committed:
[fio toup] aktualizr-lite: Handle reboot detection and rollbacks

Add support to handle reboot detection and rollback in finalizeIfNeeded.
Notification process also changed in order to track the right state
depending on pending, rollback or current.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ricardosalveti committed:
[fio toup] aktualizr-lite/dockerapp: avoid multiple current ostree installs

Compare current against target OSTree hash and skip the install call if
the same. Also update aktualizr-lite to handle the updated return code
from docker-app install.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ricardosalveti committed:
[fio extras] aktualizr-lite: daemon: skip older known images (rollback)

Check if the target image is known when running in daemon mode and skip
the update process if so. This is done as a way to detect rollbacks
after applying an update as the image will be stored locally in the
installation log table.

The correct implementation requires api/storage changes in order to
support broken/bad states in the installation log table, which needs to
be done in upstream first (and then drop/update this patch).

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ricardosalveti committed:
[fio toup] dockerappmanager: notify bootloader before ostree install

OStree is known to be running on the primary ECU, so notify the
bootloader before performing the install process for rollback support.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

doanac commented on src/aktualizr_lite/helpers.cc:
```diff
@@ -88,21 +106,22 @@ LiteClient::LiteClient(Config &config_in) : config(std::move(config_in)), primar
+
+  std::pair<Uptane::Target, data::ResultCode::Numeric> pair = finalizeIfNeeded(*package_manager, *storage, config.pacman);
+  headers.emplace_back("x-ats-target: " + pair.first.filename());
+  headers.emplace_back("x-ats-tags: " + boost::algorithm::join(config.pacman.tags, ","));
```
The headers need to be set *before* we instantiate the `HttpClient` or else they won't get set.

ricardosalveti replied:
Hm, what I was trying here is to run finalizeIfNeeded after setBootOK, which gets called via makePackageManager, so we can have the right call order.

As we need the headers before instantiating httpclient (and pair.first.filename), I guess there is no easy way to avoid finalizing before setting boot ok, but I guess this should be fine still.

doanac replied:
I think the httpclient now has an "update-header" type API that i got upstream. So you could use that if it makes things cleaner.

doanac commented on src/libaktualizr/package_manager/dockerappmanager.cc:
```diff
@@ -142,7 +142,18 @@ bool DockerAppManager::fetchTarget(const Uptane::Target &target, Uptane::Fetcher
+  } else {
+    LOG_INFO << "Target " << target.sha256Hash() << " is same as current";
+    res = data::InstallationResult(data::ResultCode::Numeric::kOk, "OSTree hash already installed, same as current");
+  }
```
Very smart - This is worth upstreaming. You want to try and submit it, or would you rather I deal with it?

ricardosalveti replied:
Probably better for you to upstream as I still need to go over all the testing requirements, which is something you are already more familiar with.

doanac commented on src/aktualizr_lite/main.cc:
```diff
@@ -239,20 +242,43 @@ static int daemon_main(LiteClient &client, const bpo::variables_map &variables_m
+      // just check if the version is known (old hash) and not current/pending and abort if so
+      bool known_target_sha = false;
+      boost::optional<Uptane::Target> pending;
+      client.storage->loadInstalledVersions("", nullptr, &pending);
```
There's a newer-ish API that's a little more clear about the intent, `loadPrimaryInstalledVersions`, you probably should use here.

ricardosalveti replied:
Sure, will change.

doanac commented on src/aktualizr_lite/main.cc:
```diff
@@ -222,6 +222,9 @@ static int daemon_main(LiteClient &client, const bpo::variables_map &variables_m
   }
 
+  std::vector<Uptane::Target> installed_versions;
+  client.storage->loadPrimaryInstallationLog(&installed_versions, false);
```
I guess this doesn't need to be in the while-loop because OSTree related changes that would show up here are going to require a reboot, so loading only needs to be done once?

ricardosalveti replied:
Yes, because the only other possibility here is pending images (which could be multiple ones, but should still be fine).

doanac commented on src/aktualizr_lite/main.cc:
```diff
@@ -239,20 +242,43 @@ static int daemon_main(LiteClient &client, const bpo::variables_map &variables_m
+              continue;
+            }
+            LOG_INFO << "Target sha256Hash " << target->sha256Hash() << " known locally (rollback?), skipping";
+            known_target_sha = true;
```
This might be problem for an edge case. Lets say we have two builds, 1 and 2 both tagged "promoted" and the device listens only accepts promoted. If you removed the "promoted" tag from build 2 and update for build 1 would get triggered here. I *think* this logic would reject the update, right? There is a `was_installed` parameter to the `loadInstalledVersions` method. I wonder if we could use that to get a list of failed installs?

doanac replied:
It would also be good to break this out into its own function. I think if we put it in helpers.cc we could even create some unit tests to cover different failure scenarios.

ricardosalveti replied:
> This might be problem for an edge case. Lets say we have two builds, 1 and 2 both tagged "promoted" and the device listens only accepts promoted. If you removed the "promoted" tag from build 2 and update for build 1 would get triggered here. I _think_ this logic would reject the update, right? There is a `was_installed` parameter to the `loadInstalledVersions` method. I wonder if we could use that to get a list of failed installs?

The problem is that every image state gets stored into installed versions, like after applying the install (pending), after confirming the boot (current, when was_installed is also set to 1) and after marking the image as bad (none). If we are able to have a separated field for 'broken' images, then we could use the install log to only retrieve the ones we want, as currently we can only retrieve ones that were set to current || everything.

ricardosalveti replied:
What we could do instead is using was_installed to retrieve the list of good images and assume it is not bad if part of the list or equal to pending, but that would basically be a variation of what this piece is already doing.

doanac commented on src/libaktualizr/package_manager/dockerappmanager.cc:
```diff
@@ -145,6 +145,11 @@ data::InstallationResult DockerAppManager::install(const Uptane::Target &target)
+    // and a false notification doesn't hurt with rollback support in place
+    if (bootloader_ != nullptr) {
+      bootloader_->updateNotify();
+    }
```
we should title this commit `[fio toup]` so we remember to push it upstream.

ricardosalveti replied:
Sure, will change.

ricardosalveti committed:
[fio extras] aktualizr-lite: helpers: add known_local_target

Function used to check if a target is known locally and protect the
system from generating an update loop after a rollback.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

doanac commented on src/aktualizr_lite/helpers.cc:
```diff
@@ -88,21 +106,22 @@ LiteClient::LiteClient(Config &config_in) : config(std::move(config_in)), primar
-  }
+  std::pair<Uptane::Target, data::ResultCode::Numeric> pair = finalizeIfNeeded(*package_manager, *storage, config.pacman);
+  http_client->updateHeader("x-ats-target", pair.first.filename());
+  http_client->updateHeader("x-ats-tags", boost::algorithm::join(config.pacman.tags, ","));
```
It looks like [`updateHeader`](https://github.com/advancedtelematic/aktualizr/blob/master/src/libaktualizr/http/httpclient.cc#L261) requires the header to already be defined. The "x-ats-tags" header can stay where it was, so the real problem you have is with the "x-ats-target" header. I can see two ways to deal with it:
* Define it above with an value like "?" and then updateHeader will work
* The calls you make to package-manager don't require a real HttpClient, so you could just pass a default constructed HttpClient just to that object

ricardosalveti replied:
Hm, looking at the code it seems http is not really used by any package manager, but I guess it could at some point?

And nice catch, I had the impression it would update or add to the list of known headers but you are right :-)

Guess I will go with the first suggestion to be more aligned with what package-manager expects.

ricardosalveti replied:
Pushed an update following your first suggestion, setting to 'unknown' first and later updating based on the output of finalizeIfNeeded.

doanac reviewed (commented):
Just one minor problem

doanac commented:
Looks good. Thanks!
